adding a slight delay for searching

tejans24 opened the pull request:
connects 1201

mdbenjam commented on client/app/reader/actions.js:
```diff
@@ -22,13 +22,27 @@ export const toggleExpandAll = () => ({
+    });
+
+    // adding a slight delay for performing search
+    _.debounce(() => {
```
I don't think adding the dbounce here will work. We're creating a closure every time we go through the function. Instead we need one _.debounce method that is repeatedly called. The debounce should be move to performSearch.

tejans24 replied:
This closure works in this case but I agree that the placement of this can be better. I'll look into modifying this.

tejans24 commented:
I had to keep track of the state of the text value in both the redux state and in the component for this to work properly. The function is debouced before the action is called.

NickHeiner commented:
Ok, once I crank `reader_id2` up to `1000` documents, I can see this making a difference. Let's go ahead and check this change in so we have a better test case:

```diff
diff --git a/lib/fakes/appeal_repository.rb b/lib/fakes/appeal_repository.rb
index b2cdf474..f2b0a5bd 100755
--- a/lib/fakes/appeal_repository.rb
+++ b/lib/fakes/appeal_repository.rb
@@ -376,7 +376,7 @@ class Fakes::AppealRepository
         veteran_first_name: "Joe",
         veteran_last_name: "Smith"
       },
-      documents: random_reader_documents(200)
+      documents: random_reader_documents(1000)
     )
     Generators::Appeal.build(
       vacols_id: "reader_id3",

```

NickHeiner commented:
This is causing a new console error to appear:

![image](https://cloud.githubusercontent.com/assets/829827/25971585/cc1d3d9a-366a-11e7-8472-981ed946598c.png)

Please check out that link in the error message and resolve the issue that they're describing.

Related: We should not be storing anything in local `state`. We can put everything in `redux`. We just need to decouple the notion of updating the search query and updating the search results. We should have two separate actions: `UPDATE_SEARCH_QUERY` and `UPDATE_SEARCH_QUERY_FROM_RESULTS`.

NickHeiner commented on client/app/reader/actions.js:
```diff
@@ -35,12 +35,16 @@ export const toggleExpandAll = () => ({
-  payload: {
-    searchQuery
+export const setSearch = (searchQuery) => (
+  (dispatch) => {
```
Why was this change necessary? If we're doing `bindActionCreators`, why do we need to explicitly reference `dispatch` here?

tejans24 commented:
btw, I tried a few iterations of having multiple actions with debounce before. 

This option doesn't work, since debouce hogs performSearch for the time given.
```
  constructor(props) {
    super(props);
    // debounce the passed in dispatch method
    this.changed = _.debounce(this.props.performSearch, DEBOUNCE_TIME_IN_MS);
  }

  handleChange = (value) => {
    this.props.setSearch(value);
    this.changed();
  }
```

Alternate is to use dispatching performSearch action after setting the search term by using debouce closure.
```
export const performSearch = () => ({
  type: Constants.PERFORM_SEARCH
});

export const setSearch = (searchQuery) => (
  (dispatch) => {
    dispatch({
      type: Constants.SET_SEARCH,
      payload: {
        searchQuery
      }
    });

    // adding a slight delay for performing search
    _.debounce(() => {
      dispatch(performSearch());
    }, DEBOUNCE_TIME_IN_MS)();
  }
);
```
The placement of the debouce doesn't seem right.

Other option is in the current PR. Any other suggestions?

NickHeiner commented:
Yes, there is another option. Something like:

```js
updateSearchResults = _.debounce(() => {
  this.props.updateSearchResults();
})

handleChange = (newSearchValue) => {
  this.props.updateSearchValue(newSearchValue);
  this.updateSearchResults();
}

render() {
  return <TextField onChange={this.handleChange} />
}

```

Obviously there's a lot of detail missing there, but hopefully this makes sense. Let me know if it doesn't; I'll be at BVA tomorrow if you want to discuss in person.

NickHeiner commented:
After an in-person discussion, we are going to close this for now. It is better to address the root cause with #1983.
